### PR TITLE
Refactor async blob read to avoid blocking calls, support non multipa…

### PR DESCRIPTION
**Pull request by kotwanikunal**
…rt calls

### Description
- Refactor async blob read to avoid blocking calls for object attributes called out as a part of https://github.com/opensearch-project/OpenSearch/pull/9710#issuecomment-1730482740
- Current requests for objects did not accommodate for objects stored without a multipart call - added support for non multipart objects
- Fixed object access for keys with a base path within the repository

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java [hunk 1]
```diff
@@ -685,41 +700,47 @@ static Tuple<Long, Long> numberOfMultiparts(final long totalSize, final long par
...
-    static InputStreamContainer transformResponseToInputStreamContainer(ResponseInputStream<GetObjectResponse> streamResponse) {
+    static InputStreamContainer transformResponseToInputStreamContainer(
+        ResponseInputStream<GetObjectResponse> streamResponse,
+        boolean isMultipartObject
+    ) {
         final GetObjectResponse getObjectResponse = streamResponse.response();
         final String contentRange = getObjectResponse.contentRange();
         final Long contentLength = getObjectResponse.contentLength();
-        if (contentRange == null || contentLength == null) {
+        if ((isMultipartObject && contentRange == null) || contentLength == null) {
```
Comment (andrross) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:740]: Is contentRange actually null in the non multipart case? I would expect it always to be present and not to need conditional logic here.

Comment (kotwanikunal) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java:740]: It is `null` in the non multipart case. I tried looking up the client contract for S3 - could not locate this specific attribute behavior. But in my tests - it is null in non multipart cases.

**Comment by kotwanikunal** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:**  ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/26144/
> * **CommitID:** [65e9ec9](https://github.com/opensearch-project/OpenSearch/commit/65e9ec98ce953db8701a48151aac588056624821)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Looks like Jenkins crashed again.

image

**Comment by kotwanikunal** [comment]
#10137 

> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/26149/
> * **CommitID:** [65e9ec9](https://github.com/opensearch-project/OpenSearch/commit/65e9ec98ce953db8701a48151aac588056624821)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

**Comment by kotwanikunal** [comment]
#10006 

> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/26153/
> * **CommitID:** [65e9ec9](https://github.com/opensearch-project/OpenSearch/commit/65e9ec98ce953db8701a48151aac588056624821)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

**Review by andrross — APPROVED** [review]

**Pull request merged**